src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+            => builder.AddClrTypeMappings(Assembly.GetCallingAssembly());
+
+        /// <summary>
+        /// Scans the specified assembly for classes that inherit from <see cref="ObjectGraphType{TSourceType}"/>,
+        /// <see cref="InputObjectGraphType{TSourceType}"/>, or <see cref="EnumerationGraphType{TEnum}"/>, and
+        /// registers clr type mappings on the schema between that class and the source type or underlying enum type.
+        /// Skips classes where the source type is <see cref="object"/>, or where the class is marked with
+        /// the <see cref="DoNotMapClrTypeAttribute"/>.
+        /// </summary>
+        /// <remarks>
+        /// This is equivalent to calling <see cref="SchemaExtensions.RegisterTypeMappings(ISchema, Assembly)"/>
+        /// within the schema constructor.
+        /// </remarks>
+        public static IGraphQLBuilder AddClrTypeMappings(this IGraphQLBuilder builder, Assembly assembly)
+        {
+            if (assembly == null)
+                throw new ArgumentNullException(nameof(assembly));
+
+            // retreive all of the type mappings ahead-of-time, in case of a scoped or transient schema,
+            // as reflection is relatively slow
+            var typeMappings = assembly.GetClrTypeMappings();
+            builder.ConfigureSchema((schema, serviceProvider) =>
+            {
+                foreach (var typeMapping in typeMappings)
+                {
+                    schema.RegisterTypeMapping(typeMapping.ClrType, typeMapping.GraphType);
+                }
+            });
+
+            return builder;
+        }
+        #endregion
+
+        #region - AddDocumentListener -
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentListener"/> with the dependency injection framework as both <typeparamref name="TDocumentListener"/> and
+        /// <see cref="IDocumentExecutionListener"/>. Configures document execution to add an instance of <typeparamref name="TDocumentListener"/> to the
+        /// list of document execution listeners within <see cref="ExecutionOptions.Listeners"/>. Singleton, scoped and transient lifetimes are supported.
+        /// </summary>
+        /// <remarks>
+        /// Do not separately add the document listener to your execution code or the document listener may be registered twice for the same execution.
+        /// </remarks>
+        public static IGraphQLBuilder AddDocumentListener<TDocumentListener>(this IGraphQLBuilder builder, ServiceLifetime serviceLifetime = ServiceLifetime.Singleton)
+            where TDocumentListener : class, IDocumentExecutionListener
+        {
+            builder.Register<TDocumentListener>(serviceLifetime);
+            builder.Register<IDocumentExecutionListener, TDocumentListener>(serviceLifetime);
+            builder.ConfigureExecution(options => options.Listeners.Add(options.RequestServices.GetRequiredService<TDocumentListener>()));
```

Comment:
Here we need to think about the design. It is possible to get a memoty leak and hang. `options.Listeners.Add` is called each `DocumentExecuter.ExecuteAsync`. Supplied `ExecutionOptions` object _may_ be shared, i.e used over and over for multiple (sequential, so no thread issues) requests. Imagine some batch processing system (scheduler) that reads queries from file system and executes them savings results into databaseor whatever. It may use the same `ExecutionOptions` instance because ... bacause clearly nothing prevents doing that. The problem is that such configuration has side effect - listeners list grows. And it is completely incomprehensible for an external observer who works with top-level `AddXXX` APIs.

Reply:
Good point.  Perhaps in v5 ExecutionOptions is always a delegate passed to ExecuteAsync.

Reply:
Something like that. Yes. To safely modify options, we **must** own them.

Reply:
#2796